src/train/train.py

```diff
@@ -0,0 +1,158 @@
+        if num_training_examples is not None:
+            assert num_training_examples >= bsize
+            seeds = sample_seeds(num_training_examples, bsize)
+            data_sampler_args["seeds"] = seeds
+            task_sampler_args["seeds"] = [s + 1 for s in seeds]
+
+        xs = data_sampler.sample_xs(
+            curriculum.n_points,
+            bsize,
+            curriculum.n_dims_truncated,
+            **data_sampler_args,
+        )
+        task = task_sampler(**task_sampler_args)
+        ys = task.evaluate(xs)
+
+        loss_func = task.get_training_metric()
+
+        loss, output = train_step(model, xs.cuda(), ys.cuda(), optimizer, loss_func)
+
+        point_wise_tags = list(range(curriculum.n_points))
+        point_wise_loss_func = task.get_metric()
+        point_wise_loss = point_wise_loss_func(output, ys.cuda()).mean(dim=0)
+
+        baseline_loss = (
+            sum(
+                max(curriculum.n_dims_truncated - ii, 0)
+                for ii in range(curriculum.n_points)
+            )
+            / curriculum.n_points
+        )
+
+        if i % args.wandb.log_every_steps == 0 and not args.test_run:
+            wandb.log(
+                {
+                    "overall_loss": loss,
+                    "excess_loss": loss / baseline_loss,
+                    "pointwise/loss": dict(
+                        zip(point_wise_tags, point_wise_loss.cpu().numpy())
+                    ),
+                    "n_points": curriculum.n_points,
+                    "n_dims": curriculum.n_dims_truncated,
+                },
+                step=i,
+            )
+
+        curriculum.update()
+
+        pbar.set_description(f"loss {loss}")
+        if i % args.training.save_every_steps == 0 and not args.test_run:
+            training_state = {
+                "model_state_dict": model.state_dict(),
+                "optimizer_state_dict": optimizer.state_dict(),
+                "train_step": i,
+            }
+            torch.save(training_state, state_path)
+
+        if (
+            args.training.keep_every_steps > 0
+            and i % args.training.keep_every_steps == 0
+            and not args.test_run
+            and i > 0
+        ):
+            torch.save(model.state_dict(), os.path.join(args.out_dir, f"model_{i}.pt"))
+
+
+def main(args):
+    if args.test_run:
+        curriculum_args = args.training.curriculum
+        curriculum_args.points.start = curriculum_args.points.end
+        curriculum_args.dims.start = curriculum_args.dims.end
+        args.training.train_steps = 100
+    else:
+        wandb.init(
+            dir=args.out_dir,
+            project=args.wandb.project,
+            entity=args.wandb.entity,
+            config=args.__dict__,
+            notes=args.wandb.notes,
+            name=args.wandb.name,
+            resume=True,
+        )
+
+    model = build_model(args.model)
+    model.cuda()
+    model.train()
```

Comment:
this should be done in the `<ContextTrainer>.train(...)` method and `model.eval()` should be called just before returning the model @riensou

Reply:
right, I think all of train.py is old and just garbage from the Garg et al codebase at this point